packages/alfa-dom/src/matches.ts

```diff
@@ -178,15 +200,62 @@ function matchesClass(element: Element, selector: ClassSelector): boolean {
 /**
  * @see https://www.w3.org/TR/selectors/#type-selectors
  */
-function matchesType(element: Element, selector: TypeSelector): boolean {
+function matchesType(
+  element: Element,
+  context: Node,
+  selector: TypeSelector,
+  options: MatchesOptions
+): boolean {
   // https://www.w3.org/TR/selectors/#the-universal-selector
   if (selector.name === "*") {
     return true;
   }
 
+  if (matchesNamespace(element, context, selector, options) === false) {
+    return false;
+  }
+
   return element.localName === selector.name;
 }
 
+/**
+ * @see https://www.w3.org/TR/selectors/#type-nmsp
+ */
+function matchesNamespace(
+  element: Element,
+  context: Node,
+  selector: TypeSelector,
+  options: MatchesOptions
+): boolean {
+  if (selector.namespace === "*" || options.namespaces === undefined) {
+    return true;
+  }
```

Comment:
In the event that `options.namespaces` is `undefined`, `matchesNamespace()` should return `true` _only if_ `selector.namespace` is also `null`.